### PR TITLE
Grammar coverage results: Add number of ReCOGS training examples (augmentations excluded) required to reach full grammar coverage with median and confidence interval for shuffled training data case in results

### PR DESCRIPTION
**Pull request by willy-b**
Draft PDF for this PR at https://raw.githubusercontent.com/willy-b/RASP-for-ReCOGS/76edf39a4471e7793d5f2d7a0f40f2cfb2d49699/rasp-for-recogs_pos-wbruns-2024-draft.pdf

Replace grammar coverage by ReCOGS training example (excluding augmentations) using the default training set order with median and 95% confidence bounds lines over n=1000 random permutations of the training set (could also do bootstrap resample but original author is randomly shuffling before training their Transformers so using shuffles).

Screenshot_20250106_091348

Recall separate from this figure or the grammar coverage arguments we built a model by hand in RASP which can theoretically be compiled to Transformer neural network weights, to prove by construction that a Transformer can perform the ReCOGS task compositionally and systematically (see 100% semantic exact match and string exact match results posted with notebooks at e.g. https://github.com/willy-b/learning-rasp/pull/7 for dev set, test set, and 100% semantic exact match on the generalization splits except the obj-pp-to-subj-pp which gets 92%). We need to show that the rules it uses could come from the training data. We show in the paper just 21 examples (19 for non-recursive parts of the grammar) are enough with lexical differences ignored to cover the ReCOGS/COGS input grammar, and those are used by the RASP model, but someone might ask whether the RASP rules (the ones after an embedding that tags words with possible part-of-speech and possible verb type) are learnable from the training set. This figure shows that even without 21 handpicked examples one will encounter the same coverage in the actual ReCOGS training set (even without augmentations) quickly just by running through in a random order (<= 161 examples with 97.5% probability, <= 77 examples with 50% probability, by 55 examples in default order; not cheating with information outside the training set or even needing to go through a large fraction of the tens of thousands of examples available) .

Regarding the RASP models initial "embedding" that adds part-of-speech and verb type information, for the actual Transformer learning from scratch, learning that would require more of the training examples. (Transformers do this: See paper for references but e.g. Tenney et al 2019 show that pre-trained BERT learns to associate part-of-speech information in the very earliest layer, layer 0 for most sentences, and in the few truly ambiguous sentences represents a good guess at an early layer and corrects it with context at a deeper layer.). But factoring the learning into learning that embedding and then learning the grammar rules shows how efficient that is in terms of representing the entire grammar  (e.g. with just 21 rules on top of an embedding which is a dictionary lookup for each word of its possible part of speech and verb types) and efficient subsequent learning of grammar rules, and was a natural (and based on others analysis of real-world Transformers trained on lots of text, realistic) way to factor the model. 

The notebook where grammar coverage with confidence intervals by ReCOGS example index for n=1000 random shuffles of the dataset is computed, is NOT shared in this case (at least yet), but the link is https://colab.research.google.com/drive/1lkjMpp2KOeZvdKO9su9ZTGG7t5JfQ4X7 . (All other notebooks for project are shared.) This is as I need to separate out Fuzzing Book related code to avoid a potential copyright issue (just to be extremely careful, though they are CC-BY-NC-SA so pretty liberal). For their Grammar Coverage code and related information, see their page already cited in the paper at https://www.fuzzingbook.org/html/GrammarCoverageFuzzer.html (or archived at https://web.archive.org/web/20241125001213/https://www.fuzzingbook.org/html/GrammarCoverageFuzzer.html ).

### CODE REVIEW

**Comment by willy-b** [comment]
Edited description to note:

Draft PDF for this PR is at https://raw.githubusercontent.com/willy-b/RASP-for-ReCOGS/76edf39a4471e7793d5f2d7a0f40f2cfb2d49699/rasp-for-recogs_pos-wbruns-2024-draft.pdf .

The notebook where grammar coverage with confidence intervals by ReCOGS example index for n=1000 random shuffles of the dataset is computed, is NOT shared in this case (at least yet), but the link is https://colab.research.google.com/drive/1lkjMpp2KOeZvdKO9su9ZTGG7t5JfQ4X7 . (All other notebooks for project are shared.) This is as I need to separate out Fuzzing Book related code to avoid a potential copyright issue (just to be extremely careful, though they are CC-BY-NC-SA https://creativecommons.org/licenses/by-nc-sa/4.0/ so pretty liberal). For their Grammar Coverage code and related information, see their page already cited in the paper (from first version) at https://www.fuzzingbook.org/html/GrammarCoverageFuzzer.html (or archived at https://web.archive.org/web/20241125001213/https://www.fuzzingbook.org/html/GrammarCoverageFuzzer.html ).
